18778 fuzzy comparisons

seadavis opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #18778 


#### Brief description of what is fixed or changed
Add to relational a function called cmp(lhs, rhs, op). Where op can be one of
In calculus utils,  and under AccumulationBounds, move the logic into _eval_is_lt, _eval_is_gt.
Did the same thing for Indexed in the tensor module.
With this new pull request, any class that inherits from Expr should implement _eval_is_X, if you want to override the default comparison. Returning None if nothing can be evaluated.

Moved the logic from Expr cmp into cmp(lhs, rhs, op).
Moved the logic from Equlaity into cmp - which calls _is_eq
The logic under _InEquality also uses the cmp logic.

 Examples:
    >>> from sympy.core.relational import cmp
    >>> from sympy import sympify
    >>> from sympy import Symbol
    >>> x = Symbol('x')
    >>> two = sympify(2)
    >>> cmp(x,two, "<") is None

    >>> from sympy.core.relational import cmp
    >>> from sympy import sympify
    >>> from sympy import Symbol
    >>> x = Symbol('x')
    >>> two = sympify(2)
    >>> cmp(x,two, "=") is None

    >>> from sympy.core.relational import cmp
    >>> from sympy import sympify
    >>> from sympy import Symbol
    >>> three = sympify(3)
    >>> two = sympify(2)
    >>> cmp(three,two, "<") is False

    >>> from sympy.core.relational import cmp
    >>> from sympy import sympify
    >>> from sympy import Symbol
    >>> three = sympify(3)
    >>> two = sympify(2)
    >>> cmp(three,two, ">") is True

#### Other comments
_eval_Eq doesn't match with _eval_is_gt however,there are already a great number of classes

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

seadavis committed:
Big Number: Comparison of large exponentials

Implemented a class that allows us to implement very
large exponents. There is still more work to do, but
this will at least allow us to compare number such
as 10**10**100. We cannot, in general,compare unevaluated
expressions yet, so multiplication addition and subtraction,
are just evaluated unless one contains some Power
But Pow's will be unevaluated and symbolically

Examples:
     >>> from sympy.core.numbers import BigNumber,
     >>>
     >>> e1 = Pow(BigNumber(10), Pow(BigNumber(10)
         BigNumber(100)))
     >>> 100 < e1
     True

     >>> BigNumber(10)*BigNumber(2)
     20

seadavis committed:
Update sympy/core/numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/tests/test_numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/tests/test_numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/tests/test_numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Implemented, Fuzzy Comparisons over numbers

relational: Implemented Fuzzy Comparison
Moved, logic from Eq to take place under a new function cmp(lhs, rhs, op),
which does a fuzzy comparison.

seadavis committed:
Update relational.py

oscarbenjamin commented:
This looks good although I need more time to look through it.

It has a few small merge conflicts.

oscarbenjamin commented:
I don't think we want `_eval_is` methods for all 6 of lt, le, eq, ne, ge, gt because there is a lot of redundancy there. Each of these is the logical negation of another so e.g.:
```
def is_ne(a, b):
    return fuzzy_not(is_eq(a, b))
```
Since ne == ~eq, lt == ~ge, le == ~gt we can eliminate 3 of the 6 reducing it to eq, ge and gt. We can provide `is_*` functions for all 6 but we can build all 6 from a lower level that only implements 3 of them.

If all comparisons were decidable we could take that down to 2 like total ordering does
https://docs.python.org/2/library/functools.html#functools.total_ordering
but in sympy we have undecidable comparisons like `x**2 > 0` (for real x). Here we can say that ge is True but eq and gt are both None.

We do however have relationships between eq, ge and gt e.g.:
```
eq -> ge
gt -> ge
eq -> ~gt
gt -> ~eq
~ge -> ~eq
~ge -> ~gt
```
So we can use those implications as part of `is_eq`, `is_ge` and `is_gt`.

Under the hood then we need three functions to delegate to like `_eval_is_eq`, `_eval_is_ge`, and `_eval_is_gt`. I think maybe these should use multiple dispatch rather than being `_eval_is` methods though. See e.g. the `is_subset` method for sets and the handlers in sympy/sets/handlers/issubset.py.

So I propose a number of levels:
```
1. Expr.__lt__, ...
2. LessThan, ...
3. is_lt, is_le, is_eq, is_ne, is_ge, is_gt
4. _is_equal, _is_less_than, _is_less_than_or_equal
```
Level 3 are the internal functions that can be used along with e.g. `x.is_positive` to control evaluation of Basic objects. Level 3 is built out of level 4 using the implication rules and equivalences.

Then level 4 can use multiple dispatch to decide how to handle e.g. Pow < Number etc. Hopefully at this level we can provide the minimum logic needed to implement all of the levels above without needing to duplicate it for too many of the different combinations of types.

oscarbenjamin commented:
> Each of these is the logical negation of another so e.g.:
> 
> ```
> def is_ne(a, b):
>     return fuzzy_not(is_eq(a, b))
> ```
> 
> Since ne == ~eq, lt == ~ge, le == ~gt

Actually this is only true if both a and b are (extended) real...

seadavis commented:
so then in the case of AccumulationBounds and in, tensor/indexed we would use multiple dispatch, instead of putting the logic in the __x__ operators, or the _eval_is_X?

oscarbenjamin commented:
> so then in the case of AccumulationBounds and in, tensor/indexed we would use multiple dispatch, instead of putting the logic in the **x** operators, or the _eval_is_X?

That is what I am suggesting. It is not the only way to make this work but it seems like a good approach to me. Multiple dispatch is designed to be scalable in the case where you want to implement a function across a Cartesian product of different input types which is the case here.

Abdullahjavednesar commented:
@seadavis there are conflicts here, can you please resolve them.

smichr commented on sympy/core/relational.py:
```diff
@@ -1131,3 +1033,269 @@ def _eval_relation(cls, lhs, rhs):
+    """
+    Resolves cmp(lhs, rhs) to S.true if cmp returns true.
+    or to S.false if cmp returns false
+    or to None if cmp returns none.
```
It looks like it returns an unevaluated `cls` via line 1058

smichr commented on sympy/core/relational.py:
```diff
@@ -1131,3 +1033,269 @@ def _eval_relation(cls, lhs, rhs):
+        else:
+            return is_eq
+
+    for me in (lhs, rhs):
```
We might want to make this return False in the first case; for `nan`, maybe None instead of False.

smichr replied:
See discussion [here](https://groups.google.com/forum/#!topic/sympy/W0iKfAGsc2E).

seadavis commented:
@Abdullahjavednesar I certainly will.
@oscarbenjamin I agree that will definitely be a nice way of handling these things. I can add to the implementation sometime this week.

seadavis committed:
relation: Reduced number of comparisons

Reduced the number of _eval_is operations necessary to override default expression comparisons.

seadavis committed:
Merge branch 'master' into 18778_FuzzyComparisons

seadavis committed:
relational: Addmethods

added is_lt, is_eq to methods.

seadavis commented:
@Abdullahjavednesar I've taken care of the merge conflicts.
@oscarbenjamin I've taken most of your suggestions into consideration. Except for the multiple dispatch, because that does cause a circular import Relational -> handlers.py -> AssumBounds -> Relational. Even if you put it in the same file you'd still have Relational -> AssumBounds -> Relational, unless there's another way to do it that I don't know about.

I also provided is_eq etc. methods, but left cmp with the op. If the sections at 1204 and 1219 can also be done with just eq, is_gt, is_ge then we can simplify this further and remove the _eval_is methods out of cmp, I just wasn't sure if is_extended_positive was always going to be the negation of is_extended_negative?

seadavis committed:
Update relational.py

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -490,121 +490,20 @@ def __new__(cls, lhs, rhs=None, **options):
 
     def _eval_rewrite_as_Add(self, *args, **kwargs):
+        from .add import _unevaluated_Add, Add
         """return Eq(L, R) as L - R. To control the evaluation of
```
There should not be code before the docstring

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1033,308 @@ def _eval_relation(cls, lhs, rhs):
+        else:
+            return S.false
+    else:
+        return val
```
Should this case not be an error? Can op return anything other than True, False or None?

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1033,308 @@ def _eval_relation(cls, lhs, rhs):
+
+def fuzzy_not(val):
+    if val is None:
+        return val
```
This is already defined in sympy/core/logic

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1033,308 @@ def _eval_relation(cls, lhs, rhs):
+    negate_op = False
+    if op == "<":
+        eval_str = eval_str + "is_ge"
+        negate_op = True
```
The "<" and ">=" operations are only mutually exclusive when both operands are guaranteed to be extended real. Is that checked somewhere else?

seadavis replied:
no, it is not. I'll fix that.

oscarbenjamin commented:
I find it strange that the `is_eq` etc. functions all call through to e.g. (`cmp(..., op='=')`) which then calls `eval_dispatch` which then dispatches on `op` to get to `_eval_is_eq`. I would expect that it could be done more directly.

The `cmp` function will then go on to call `_cmp_eq`. Shouldn't the code in `_cmp_eq` just be part of `is_eq`?

seadavis commented:
> I find it strange that the `is_eq` etc. functions all call through to e.g. (`cmp(..., op='=')`) which then calls `eval_dispatch` which then dispatches on `op` to get to `_eval_is_eq`. I would expect that it could be done more directly.

This is true, actually, I didn't think it could yesterday, but I see a solution now.

> The `cmp` function will then go on to call `_cmp_eq`. Shouldn't the code in `_cmp_eq` just be part of `is_eq`?
This is true, I'll make that refactor.

seadavis committed:
relational : Travis fix & direct calling of _eval_is_X methods

Refactored the fuzzy comparisons such that, the _eval_is_X is being called more directl, and to fix the travis builds.

seadavis committed:
relational: Fix travis errors

oscarbenjamin commented:
I find the logic here really confusing between `is_le`, `_eval_cmp`, `_eval_dispatch`, `cmp`. There must be a way of simplifying all of that and making the calls more direct. We seem to funnel everything into one function and then dispatch back out again in multiple places. I presume that the reason for this DRY but I'd rather have a bit of repetition than code this confusing.

seadavis commented:
I know what you mean. The reason was DRY, but I debated whether it was better that way or to have some repetition. I'll remove the _eval_cmp function, and basically get all functions to try _eval_dispatch first, and if None then to try cmp

seadavis committed:
relational: called _eval_dispatch more directly

seadavis commented:
@oscarbenjamin, so if you want to eliminate _eval_dispatch, I can do that and just put the call to evaluate inside of the is_x methods.But there will be repeated code, and my other concern is that it removes a certain amount of consistency - check _eval_is_X, if return None, then use the standard sympy comparison. Which does make it easy to write objects that use that API, Is there perhaps an super-object that all the object we care about are using that  I should use to encapsulate that logic.

oscarbenjamin commented:
There would be a bit of repetition but the code would be easier to understand and to make correct. I'm not convinced that the current logic is correct for example we have:
```python
def is_gt(lhs, rhs):
    retval = _eval_dispatch(lhs, rhs, "_eval_is_gt")
    if retval is None:
        retval = cmp(lhs, rhs, ">")
    return retval

def _eval_dispatch(lhs, rhs, eval_str):
    retval = None
    if hasattr(lhs, eval_str):
        retval = getattr(lhs, eval_str)(rhs)
    elif hasattr(rhs, eval_str):
        retval = getattr(rhs, eval_str)(lhs)

    return retval
```
It's quite confusing but if I read this correctly then `is_gt(lhs, rhs)` will return `lhs._eval_is_gt(rhs)` if lhs has the method or otherwise `rhs._eval_is_gt(rhs)` if rhs has the method. That can't make sense though because it should surely be `rhs._eval_is_le(lhs)`. It would be easier to see whether this code is correct or not if the extra level of indirection (`_eval_dispatch`) was not there.

seadavis commented:
@oscarbenjamin I do see your point. I'll make those corrections.

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1407,47 +1404,11 @@ def __lt__(self, other):
             if self.min >= other:
                 return False
-        return super().__lt__(other)
+        return None
```
```suggestion
```

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1407,47 +1404,11 @@ def __lt__(self, other):
-
-    def __gt__(self, other):
-        """
+        """""
```
```suggestion
        """
```

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1481,9 +1442,10 @@ def __gt__(self, other):
             if self.max <= other:
                 return False
-        return super().__gt__(other)
+        return None
```
```suggestion
```

czgdp1807 commented on sympy/calculus/util.py:
```diff
@@ -1518,7 +1479,8 @@ def __ge__(self, other):
             if self.max < other:
                 return False
-        return super().__ge__(other)
+        return None
```
```suggestion
```

czgdp1807 commented on sympy/core/expr.py:
```diff
@@ -1,5 +1,7 @@
+from __future__ import print_function, division
 from typing import Tuple as tTuple
 
+
```
```suggestion
```

czgdp1807 commented on sympy/core/expr.py:
```diff
@@ -3862,24 +3820,14 @@ def __new__(cls, arg, **kwargs):
         obj = Expr.__new__(cls, arg, **kwargs)
         return obj
 
+
```
```suggestion
```

czgdp1807 commented on sympy/core/expr.py:
```diff
@@ -3961,6 +3909,8 @@ def search_element(self, elem):
         return None
 
 
+
+
```
```suggestion
```

czgdp1807 commented on sympy/core/numbers.py:
```diff
@@ -3928,6 +3929,10 @@ def sympify_complex(a):
 
 converter[complex] = sympify_complex
 
+
+
+
+
 from .power import Pow, integer_nthroot
```
```suggestion
from .power import Pow, integer_nthroot
```

czgdp1807 commented on sympy/core/tests/test_numbers.py:
```diff
@@ -1,3 +1,6 @@
+
+from __future__ import absolute_import
+
```
```suggestion
```
Python 3 is only supported so `__future__` not needed.

czgdp1807 commented on sympy/core/tests/test_relational.py:
```diff
@@ -690,7 +690,7 @@ def test_issue_8245():
     assert rel_check(a, r)
 
-    assert Eq(log(cos(2)**2 + sin(2)**2), 0) == True
+    assert Eq(log(cos(2)**2 + sin(2)**2), 0) == S.true
```
```suggestion
    assert Eq(log(cos(2)**2 + sin(2)**2), 0) is S.true
```

czgdp1807 commented:
Please remove, `return None`, the function automatically returns `None` in python if it reaches it's end.

seadavis committed:
Update sympy/core/tests/test_relational.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/tests/test_numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/numbers.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/expr.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/expr.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/core/expr.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/calculus/util.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/calculus/util.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/calculus/util.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
Update sympy/calculus/util.py

Co-Authored-By: Gagandeep Singh <gdp.1807@gmail.com>

seadavis committed:
relationa: made is_le, etc. Eliminated _eval_dispatch

smichr commented on sympy/core/expr.py:
```diff
@@ -23,6 +24,10 @@ class Expr(Basic, EvalfMixin):
     used only for argument storage and expression manipulation, i.e.
     pattern matching, substitutions, etc).
 
+    Should us _eval_is_X (_eval_is_gt/_eval_is_lt _eval_is_le) if we want to override
```
us -> use and is ge left off on purpose? perhaps parenthetical should read "(where X is gt, lt, ge, or le)"

seadavis replied:
Ah, yes you are correct thank you.

seadavis committed:
relational: removed explicitly returning None. Fixed spelling errors in Expr.py

smichr commented on sympy/core/expr.py:
```diff
@@ -24,8 +24,8 @@ class Expr(Basic, EvalfMixin):
-    Should us _eval_is_X (_eval_is_gt/_eval_is_lt _eval_is_le) if we want to override
-    the comparison functions.
+    If you want to override the comparisons of expressions:
+    Should use _eval_is_X (where X is gt, ge, lt, or Eq).
```
Do you mean `le` or `Eq`?

seadavis replied:
No I don't. We do not check for le, only for less than, and that's only the case when both expressions have, is_extended_real set to true. Less than or equal to, is just fuzzy_not over greater than.

smichr replied:
Isn't `ge` fuzzy_not over less than? (or were you setting me up to ask this :-))

seadavis replied:
Well, it might be. Sorry, I made a mistake in what I wrote last night. I meant we only check for less than when one or both expressions have is_extended_real set to false. If both number are not extended reals, ge may not be fuzzy_not over less than. See oscarbenjamin's earlier comments.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1032,265 @@ def _eval_relation(cls, lhs, rhs):
+def is_neq(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_Eq"):
+        retval = fuzzy_not(getattr(lhs, "_eval_Eq")(rhs))
```
Can this whole function just be `return fuzzy_not(is_eq(lhs, rhs))`?

seadavis replied:
yes

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1032,265 @@ def _eval_relation(cls, lhs, rhs):
+def is_eq(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_Eq"):
+        retval = getattr(lhs, "_eval_Eq")(rhs)
```
Why is this using `getattr`? Can it not just be:
```
retval = lhs._eval_Eq(rhs)
```

seadavis replied:
What if _eval_Eq doesn't exist on lhs?

seadavis replied:
Sometimes _eval_Eq doesn't exist on the object. And since we're not really using inheritance for this, if i try to call _eval_Eq on an object that does not exist it will throw an error.

oscarbenjamin replied:
I mean specifically this line. If we've already checked with `hasattr` there's no need to use `getattr`. The two arg form of `getattr` raises if the attribute is not found anyway.

We could just add the `_eval_Eq` method on `Basic` and have it return `None`.

I think it's fine to document that these functions are internal and only for `Basic` and that they don't sympify.

seadavis replied:
oh, just looked up how hasattr works, I see what you're saying now. And actually implementing _eval_is_X on Basic objects, really does make sense to me. Since everything will be sympified by this point. So, if there's no problem with that, that will be the approach I will take.

oscarbenjamin commented:
I don't see the purpose of the `cmp` or `fuzzy_resolve` functions.

seadavis commented:
@oscarbenjamin, the purpose of fuzzy_resolve was to reduce the amount of code reuse. Although based on your earlier comments I will guess that you want me to again, have some degree of repeition for simplicity. This can be done. 

The purpose of cmp was to put the logic from Expr into one place, to make this change much easier to handle.

seadavis committed:
eliminated _cmp, added _eval_is_X to basic.

Added _eval_is_X to the Basic object. Eliminated _cmp and fuzzy_resolve,fromthe comparisons. Moved the logic from _cmp into is_gt, is_ge and is_lt.is_Eq holds the logic from _cmp_eq

seadavis commented:
@oscarbenjamin ok, I'll double check to see that is_lt, always returns None, and if it does I'll simplify the function.

seadavis committed:
reduced is_lt to just be fuzzy not over is_ge, removed is_lt override in AccumBounds.

oscarbenjamin commented:
Sorry @seadavis I missed the fact that there were new changes. I'll look it this ASAP

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1808,6 +1808,34 @@ def _exec_constructor_postprocessors(cls, obj):
         return obj
 
 
+    def _eval_is_lt(self, other):
```
I don't think that this method is needed any more

oscarbenjamin commented:
I wonder if `_eval_is_gt` is needed. Python's total order decorator can get away with just one inequality and eq as the minimal operators:
https://docs.python.org/3.8/library/functools.html#functools.total_ordering

If we had ge and eq then:
```
eq == eq
ne == !eq
ge == ge
gt == ge & !eq
lt == !ge
le == !ge | eq
```
So we define all 6 in terms of two primitive functions.

The problem though is `le == !ge | eq` which uses OR. That potentially does not work in fuzzy logic because we might have a situation where le is knowable but ge and eq aren't. For example:
```julia
In [5]: 0 <= x**2                                                                                                                 
Out[5]: True

In [6]: 0 < x**2                                                                                                                  
Out[6]: 
 2    
x  > 0

In [7]: Eq(0, x**2)                                                                                                               
Out[7]: 
     2
0 = x 
```
In this situation we would not be able to infer `le` from `!ge` (i.e. `lt`) and `eq` because they are not knowable separately.

I guess that means that we do need at least 3. The question then is what combinations make sense. For example we could have `le, ge, eq` or we could have `ge, gt, eq`. All possibilities (we know we need `eq` because it also applies to the non-real case):
```
1. ge, gt, eq
2. le, ge, eq
3. lt, gt, eq
```
I think the other possibilities are redundant under symmetry or less compelling:
```
4. le, lt, eq
5. le, gt, eq
6. lt, ge, eq
```
So currently we have 1 in this PR. I think that 3 does not work because if we only have lt, gt and eq then le and ge have to be expressed using OR (see above). The only possibilities then are 1 or 2.

Comparing 1 (ge, gt, eq) and 2 (le, ge, eq) I see that 2 has a nice symmetry to it since we only ever need to determine non-strict inequalities and `le(a, b)` is equivalent to `ge(b, a)` so much of the code can be shared just by swapping arguments. The inference table would then be:
```
eq == eq
ne == !eq
ge == ge
le == le
gt == ge & !eq
lt == le & !eq
```
So maybe it makes sense to define everything in terms of the primitive functions `_eval_is_le`, `_eval_is_ge` and `_eval_Eq`.

oscarbenjamin commented:
You would think that the symmetry `le(a, b) == ge(b, a)` would allow eliminating one but the problem is if these are methods then we don't necessarily have `a._eval_is_le(b) == b._eval_is_ge(a)` unless both the classes of `a` and `b` implement all of the same logic. This is why multiple dispatch would be useful here rather than using methods.

oscarbenjamin commented:
Also the `_eval_is_le` etc methods should be implemented on `Expr` rather than `Basic`. Only `_eval_Eq` can make sense on `Basic`.

seadavis commented:
I can move, the _eval_is_x methods into Expr. That actually makes a lot of sense to me. As does your symmetry argument, for moving to 2 instead of 1. The problem with implementing multiple dispatch is that, classes like AccumBounds would have to import Expr (or import a file that imports Expr) then the file containing the multiple dispatch would have to import AccumBounds, which causes a circular reference. If you know of a way around that limitation I'm open to it.

oscarbenjamin commented:
The idea with multiple dispatch is that we can put the overloaded methods in different files.

I don't know what's higher in the import chain out of add, power, mul and relational but I guess the idea would be to define the master `_is_less_than_or_equal` function in the relational module and then overload that for relevant combinations of Add, Pow, Mul in relational.py. Then overloads for AccumBounds would be defined in the same file as the AccumBounds class which is downstream in the import chain.

I'm fairly sure that multiple dispatch is the way to go here with one caveat: I don't know if it would slow some things down. Inequalities with simple numbers need to be fast.

If we didn't use multiple dispatch then I think that we could get by with `_eval_is_le` and `_eval_is_ge` but it means potentially a lot of duplication across many different classes. With or without multiple dispatch there is something of a cartesian product problem that that there are `n**2` combinations of `n` classes that you might potentially want to implement handlers for so if we can minimise duplication in as many other ways as possible then that would be good.

seadavis commented:
Ah, ok. Then yeah if we can put the overloaded methods in different files, multiple dispatch makes a quite a bit more sense.

seadavis commented:
@oscarbenjamin when I use @dispatch(nan, nan) or (S.NaN, S.NaN) over _eval_Eq it ends up in a circular reference, due to some printing. Is it an acceptable solution to just go through the printing modules and move those imports that are causing issues to be local to the method, or am I missing something. Just using @dispatch(NaN, NaN) misses the cases when nan is used.

oscarbenjamin commented:
It would be better delay the imports to printing I think. What is it that imports printing when you use nan?

seadavis commented:
@oscarbenjamin, ok I can go through them and eliminate them 
When I try to import sympy. It looks like dispatch imports basic.py,and basic.py imports sstr

Here's the error message I get:
File "<stdin>", line 1, in <module>
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\__init__.py", line 51, in <module>
    from .core import (sympify, SympifyError, cacheit, Basic, Atom,
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\core\__init__.py", line 8, in <module>
    from .expr import Expr, AtomicExpr, UnevaluatedExpr
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\core\expr.py", line 3910, in <module>
    from .mul import Mul
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\core\mul.py", line 1915, in <module>
    from .numbers import Rational
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\core\numbers.py", line 3281, in <module>
    @dispatch(nan, nan)
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\multipledispatch\core.py", line 73, in _
    dispatcher.add(types, func, on_ambiguity=on_ambiguity)
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\multipledispatch\dispatcher.py", line 169, in add
    else str(c) for c in signature)
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\multipledispatch\dispatcher.py", line 169, in <genexpr>
    else str(c) for c in signature)
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\core\basic.py", line 411, in __str__
    from sympy.printing import sstr
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\printing\__init__.py", line 3, in <module>
    from .pretty import pager_print, pretty, pretty_print, pprint, pprint_use_unicode, pprint_try_use_unicode
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\printing\pretty\__init__.py", line 3, in <module>
    from .pretty import (pretty, pretty_print, pprint, pprint_use_unicode,
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\printing\pretty\pretty.py", line 13, in <module>
    from sympy.printing.conventions import requires_partial
  File "D:\SourceCode\GitHubSideProjects\sympy_fork\sympy\sympy\printing\conventions.py", line 10, in <module>
    from sympy import Derivative

seadavis commented:
@oscarbenjamin after peeking into this, I can still do multi dispatch but only if, we work around the NaN issue, as core calls dispatch which calls print which calls core. The workaround was to maybe just put this into equality.

But, we run into the same issue with Tuple. 
Essentially the issue is that, when we use dispatch on core, and then dispatch relies on printing which relies on core.

So, I think I do have to implement overloading the various classes.

oscarbenjamin commented:
Why is it only certain types that lead to the printing issue?

seadavis commented:
It could be I shouldn't even really be using dispatch on S.NaN (which is the Singleton version), look at the type renaming (nan on line 3279 in numbers.py). But if I try to just do NaN then it actually misses the case where the Singleton version is used. It seems that dispatch doesn't quite work over multiple inheritance perhaps? Why it casues import uses when using on the singleton I don't know and I'm uncertain how to find that out.

oscarbenjamin commented:
It should be a type e.g. `type(S.NaN)`.

seadavis commented:
ah, thank you! I'll give that a shot

seadavis commented:
@oscarbenjamin should Lt(nn, 0) where nn is a symbol, with the argument nonnegative=True, evaluate to false or be unevaluated? before this PR it evaluated to false but, the inequality comparison for less than used to check,  is_extended_nonpositive, after subtracting 0 and nn, and now since Lt is made up of is_le & !is_eq Lt only checks for is_extended_negative after subtracting 0 and nn. Do we need to adjust the subtraction so that is_extended_nonpositive is set to false (right now its set to None), after the difference between nn and zero is taken?

oscarbenjamin commented:
I don't know. It's hard for me to comment on these ideas without being able to see the changes that you have made.

oscarbenjamin commented:
> should Lt(nn, 0) where nn is a symbol, with the argument nonnegative=True, evaluate to false or be unevaluated?

It should evaluate to False.

seadavis committed:
Committing to get someone to take a peek at what's going on.

seadavis commented:
@oscarbenjamin sorry about that I have added a new commit to this branch. Please take a peek, and let me know your thoughts. The line I wastalking about was 1117, in is_le, this section:
`diff = lhs - rhs
            if diff is not S.NaN:
                rv = diff.is_extended_nonpositive
                if rv is not None:
                    return rv`

The test that's failing is, line 275 under core/tests/test_relational.py. This is failing because, diff.is_extended_nonpositive, used to be diff.is_extended_negative

oscarbenjamin commented:
When you say "this branch" I imagine you mean this pull request but there have been no changes pushed her for some time.

Where would I see the changes?

seadavis commented:
Crap, forgot to finish pushing them out. Sorry @oscarbenjamin.

oscarbenjamin committed:
Update basic.py

oscarbenjamin committed:
feat(core): Add comparison functions

Fix flake8 warnings from multiple dispatch.

oscarbenjamin committed:
feat(core): Add comparison functions

Fix quality complaints about newlines at the end of the file.

oscarbenjamin committed:
feat(core): Add comparison functions

Fix flake8 warnings from multiple dispatch.

oscarbenjamin commented:
I've pushed a few commits to get the tests to run. I'll try to find time to look at this soon...

oscarbenjamin commented:
Okay I see what you mean.

I also said that it is better to have AND rather than OR when dealing with fuzzy logic and I said that the inference table could be:
```
eq == eq
ne == !eq
ge == ge
le == le
gt == ge & !eq
lt == le & !eq
```
(I realise now looking back at that that I didn't mean to have ge on the rhs but maybe it does need to be there actually...)

You've presented the example of:
```julia
In [5]: nn = Symbol('nn', nonnegative=True)                                                                                       

In [6]: nn < 0                                                                                                                    
Out[6]: nn < 0
```
According to the table we resolve this with `lt == le & !eq` but le and eq both give None. The problem is that what we actually want to prove is not `lt` but rather `!lt` and for that negating the table entry gives `!lt == !le | eq` which is precisely the OR problem I showed because `!le` and `eq` are both None but `!lt` is True.

Okay so we need a full table that shows how to calculate the negations as well. If we had eq, ge and le then we have:
```
eq == eq
ne == !eq
ge == ge
le == le
gt == !le
lt == !ge
```
(If either x or y is non-real then the result of ge, le, gt and lt should be None)

Then nothing is computed using any AND/OR for the statements or their negations. Also `ge(x, y) == le(y, x)` so there should be some way to share code between ge and le.

Does that make sense?

seadavis commented:
Yes it does make sense. Thank, you. I'll make those fixes sometime next week.

seadavis committed:
Implemented multiple dispatch.

seadavis committed:
Implementing Multiple Dispach

seadavis committed:
Merge branch 'master' into 18778_FuzzyComparisons

seadavis committed:
Update immutable.py

seadavis committed:
Merge branch '18778_FuzzyComparisons' of https://github.com/seadavis/sympy into 18778_FuzzyComparisons

oscarbenjamin commented:
There is a problem with flake8 and multiple dispatch: https://github.com/sympy/sympy/issues/17795

seadavis committed:
Added nofqa for spurious flake8 warning

oscarbenjamin commented:
There are a few more flake8 errors. Note that you can run flake8 yourself locally as
```
$ flake8 sympy
```
For a specific module you can do
```
$ flake8 sympy/core/relational.py
```
which is a lot faster.

oscarbenjamin committed:
feat(core): Add fuzzy comparison functions

Add noqa to silence spurious flake8 warnings with multipledispatch.

oscarbenjamin commented:
Good work!

I just pushed a commit on here to fix the flake8 problem and get the tests running. This is looking broadly good. There is one test failure but that doesn't look too difficult.

I have some smaller scale suggestions but I'll put them in a detailed review soon (in a few days).

seadavis committed:
Fixing, Indexed errors.

seadavis committed:
Merge branch '18778_FuzzyComparisons' of https://github.com/seadavis/sympy into 18778_FuzzyComparisons

seadavis commented:
Thanks @oscarbenjamin I did fix that error you mentioned just now, you were correct it was not difficult.

oscarbenjamin committed:
feat(core): Add fuzzy comparison functions

Remove unneeded imports and blank lines

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1627,5 +1481,103 @@ def union(self, other):
+        return False
+
+@dispatch(AccumulationBounds, Basic) # type: ignore # noqa:F811
+def _eval_is_le(lhs, rhs): # noqa: F811
```
I don't see why we need to allow `Basic` rather than `Expr` for inequalities.

Using `<` currently gives a TypeError in that situation:
```julia
In [1]: Expr() < Basic()                                                                                                                       
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-4d9c8665d934> in <module>
----> 1 Expr() < Basic()

TypeError: '<' not supported between instances of 'Expr' and 'Basic'

In [2]: Expr() < Expr()                                                                                                                        
Out[2]: Expr() < Expr()
```

oscarbenjamin replied:
Actually I think it's important that `is_le` etc only work for `Expr`. We should ensure that there are no handlers for `Basic` so that `is_le(Basic(), Basic())` gives a `TypeError`.

seadavis replied:
Ok, I'll put that in and add a unit test for is_le(Basic(), Basic()) throwing a TypeError

oscarbenjamin commented on sympy/calculus/util.py:
```diff
@@ -1627,5 +1481,103 @@ def union(self, other):
+        if lhs.max <= rhs:
+            return True
+        if lhs.min > rhs:
+            return False
```
What exactly does the use of inequalities here do? Would it make more sense to use `is_le`?

seadavis replied:
Actually yeah it would make more sense, to use is_le on these functions

oscarbenjamin commented on sympy/combinatorics/permutations.py:
```diff
@@ -950,15 +950,7 @@ def __new__(cls, *args, **kwargs):
 
-        return as_Boolean(self._array_form == other._array_form)
 
     @classmethod
```
There should only be one blank line between these methods I think

oscarbenjamin commented on sympy/combinatorics/permutations.py:
```diff
@@ -3020,3 +3012,10 @@ def __new__(cls, perm, x, evaluate=None):
+def _eval_Eq(lhs, rhs):
+    if lhs._size != rhs._size:
+        return None
+    return as_Boolean(lhs._array_form == rhs._array_form)
```
I wonder if `_eval_Eq` should return a fuzzy bool rather than a `Boolean`...

seadavis replied:
that makes sense to me

oscarbenjamin commented on sympy/core/containers.py:
```diff
@@ -141,25 +141,27 @@ def index(self, value, start=None, stop=None):
-        elif r is False:
-            return S.false
+    if rhs.is_Symbol or isinstance(rhs, AppliedUndef):
+        return None
```
Maybe it would make sense to do the type checking in dispatch by having one handler for `Symbol` and one for `AppliedUndef` and then another for `Tuple`.

seadavis replied:
Is every object with is_Symbol guaranteed to be of type Symbol? I think there is one case, I can find out more details for you, but if I do find that there is class that does not subclass Symbol but has is_Symbol set to true should I edit to subclass Symbol

oscarbenjamin replied:
I'm not sure. What is the example?

seadavis replied:
nevermind, it did actually work.

oscarbenjamin commented on sympy/core/containers.py:
```diff
@@ -141,25 +141,27 @@ def index(self, value, start=None, stop=None):
+        return S.false
 
-converter[tuple] = lambda tup: Tuple(*tup)
+    r = fuzzy_and(fuzzy_bool(Eq(s, o)) for s, o in zip(lhs, rhs))
```
Should this use `is_eq` rather than `Eq`?

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -1,3 +1,4 @@
+from __future__ import print_function, division
```
We don't need future imports for py3 things as python 2 isn't supported any more.

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -23,6 +24,10 @@ class Expr(Basic, EvalfMixin):
     pattern matching, substitutions, etc).
 
+    If you want to override the comparisons of expressions:
+    Should use _eval_is_X (where X is gt, ge, lt, or Eq).
```
I guess that this list needs updating

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -339,78 +344,30 @@ def __complex__(self):
-        from sympy import GreaterThan
-        return self._cmp(other, ">=", GreaterThan)
+        from .relational import GreaterThan
+        return self._cmp(other, GreaterThan)
```
If the purpose of the `_cmp` method is just to sympy the argument can we use `sympify_return` instead?

That's what's used for e.g. `Expr.__add__`.

oscarbenjamin commented on sympy/core/numbers.py:
```diff
@@ -276,6 +276,8 @@ def _igcd2_python(a, b):
 
 # Use Lehmer's algorithm only for very large numbers.
+# The limit could be different on Python 2.7 and 3.x.
+# If so, then this could be defined in compatibility.py.
```
I think this is a hangover from a merge. These comments were probably removed because Python 2 is no longer supported.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -790,16 +699,17 @@ def __new__(cls, lhs, rhs, **options):
+        return cls._eval_fuzzy_relation(lhs, rhs)
+
+
+
```
Check the spacing in this file. Normally in sympy there are two blank lines between top-level things (e.g. classes) and one between nested items (e.g. methods)

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -790,16 +699,17 @@ def __new__(cls, lhs, rhs, **options):
 
+    @classmethod
+    def _eval_relation(cls, lhs, rhs, **options):
+        return cls._eval_fuzzy_relation(lhs, rhs)
```
I'm not sure I understand the purpose of this method. Is it there for backward compatibility?

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1066,9 +976,12 @@ class GreaterThan(_Greater):
-        return _sympify(lhs.__ge__(rhs))
+    def _eval_fuzzy_relation(cls, lhs, rhs):
+        val = is_ge(lhs, rhs)
+        if val is None:
```
There are a few of these `_eval_fuzzy_relation` methods and they differ only in which function they call (here `is_ge` for evaluating). Perhaps it would make more sense to move the other logic to the superclass `_eval_relation` method keeping only the call to `is_ge` here since that's the only bit that it unique to this class...

seadavis replied:
Yes, I agree that would be quite a bite better

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1060,169 @@ def _eval_relation(cls, lhs, rhs):
+    return None
+
+
+
```
The other functions have two spaces between them but here there are three.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1060,169 @@ def _eval_relation(cls, lhs, rhs):
+
+def is_neq(lhs, rhs):
+    from .logic import fuzzy_not
+    return fuzzy_not(is_eq(lhs, rhs))
```
I think `fuzzy_not` is already imported.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1060,169 @@ def _eval_relation(cls, lhs, rhs):
+
+    retval = _eval_Eq(lhs, rhs)
+    if retval is not None:
+        return retval
```
I think this method will still need to check e.g. `lhs._eval_Eq` for backward compatibility. Users and downstream libraries have probably defined `_eval_Eq` methods and we need to try to ensure that their code would still work with these changes.

oscarbenjamin commented on sympy/functions/elementary/integers.py:
```diff
@@ -216,6 +217,12 @@ def __lt__(self, other):
         return Lt(self, other, evaluate=False)
 
+@dispatch(floor, Basic)
+def _eval_Eq(lhs, rhs): # noqa:F811
```
I don't think a `floor` could ever compare equal to a non-`Expr` so this should probably by `(floor, Expr)`.

oscarbenjamin commented on sympy/matrices/immutable.py:
```diff
@@ -199,5 +185,27 @@ def as_immutable(self):
+
+@dispatch(ImmutableDenseMatrix, MatrixExpr)
+def _eval_Eq(lhs, rhs): # noqa:F811
+    return None
```
Is this handler needed? Can it not fall back to a general `Basic, Basic` handler that returns None?

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -2516,3 +2485,46 @@ def set_pow(x, y):
+
+@dispatch(Interval, FiniteSet)
+def _eval_Eq(lhs, rhs): # noqa: F811
+    return false
```
Rather than have these things here I think they should go in a module in `sympy/sets/handlers/`

oscarbenjamin commented:
I think that since `_eval_Eq` is a new function (in multiple dispatch form) we can rename it and it would be better if it was called `_eval_is_eq` for consistency with `_eval_is_ge`.

Also I think that `_eval_is_eq` should return fuzzy bool (True, False, None) rather than symbolic Boolean (S.true, S.false, none...).

We also need to ensure that `is_eq` (and `Eq`) still call through to an `_eval_Eq` method if either lhs or rhs defines it. We need this for backward compatibility because users and downstream libraries will have defined their own `_eval_Eq` methods and we should ensure that they continue to work. It's premature to deprecate having an `_eval_Eq` method but there should be comments in the code where it is called explaining that the call is for backward compatibility and the method should be deprecated in future.

oscarbenjamin commented:
There should be a way to avoid calling `_eval_Eq` twice when the same handler is invoked each time. For example
```
Eq(ImmutableDenseMatrix([[x]]), ImmutableDenseMatrix([[y]])) 
```
calls the same `_eval_Eq` handler both times and gets `None` each time.

seadavis commented:
> Also I think that `_eval_is_eq` should return fuzzy bool (True, False, None) rather than symbolic Boolean (S.true, S.false, none...).

Does this mean, we need to ensure is_eq also returns a fuzzy bool, rather than a symbolic boolean?

oscarbenjamin commented:
> Does this mean, we need to ensure is_eq also returns a fuzzy bool, rather than a symbolic boolean?

Does it not already? Yes, I think it would be good if `is_eq` and `is_le` etc consistently returned fuzzy bool. These functions should all operate at a lower layer than Basic.

seadavis committed:
Eliminated Basic from _eval_is_le and _eval_is_ge

seadavis committed:
Merge branch '18778_FuzzyComparisons' of https://github.com/seadavis/sympy into 18778_FuzzyComparisons

seadavis committed:
Fixed, unit tests, made util use is_ge

seadavis committed:
Fixing blank lines

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1105,7 +1118,11 @@ def is_le(lhs, rhs):
 
 def is_ge(lhs, rhs):
+    if type(lhs) == Basic or type(rhs) == Basic:
+        raise TypeError("Cannnot compare inequality on Basic")
```
This should probably be
```
    if not (isinstance(lhs, Expr) and isinstance(rhs, Expr)):
        raise TypeError("Can only compare inequalities with Expr")
```
It's very unusual to have an object whose type is literally Basic.

seadavis committed:
Fixed raise to not use Basic and fixed permuatations to use boolean instead of sympy Boolean

seadavis committed:
Moving Eq to is_eq in containers

seadavis committed:
Fixed comments, and ast unit test

seadavis committed:
Update numbers.py

seadavis committed:
Update relational.py

seadavis committed:
Update relational.py

seadavis committed:
Update expr.py

seadavis committed:
Update relational.py

seadavis committed:
Update immutable.py

seadavis committed:
changed _eval_Eq to _eval_is_eq

seadavis committed:
changed _eval_Eq to return booleans, updated _eval_is_eq and _eval_is_ge to use booleans only

seadavis commented:
@oscarbenjamin I have made your suggested changes, and this is now ready for another review, when you get the chance.

oscarbenjamin commented:
Seems to be a teset failure in the optional dependency tests:
```
________ sympy/codegen/tests/test_applications.py:test_copying_function ________
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.7.dev0-py3.6.egg/sympy/testing/runtests.py", line 1339, in _timeout
    function()
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.7.dev0-py3.6.egg/sympy/codegen/tests/test_applications.py", line 51, in test_copying_function
    mod, info = _render_compile_import(_mk_func1(), build_dir=folder)
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/sympy-1.7.dev0-py3.6.egg/sympy/codegen/tests/test_applications.py", line 21, in _mk_func1
    whl = While(i<n, [Assignment(out[i], inp[i]), PreIncrement(i)])
TypeError: '<' not supported between instances of 'Variable' and 'Variable'
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1049,204 @@ def _eval_relation(cls, lhs, rhs):
+
+@dispatch(Expr, Expr)
+def _eval_is_le(lhs, rhs):
+    return None
```
Is `_eval_is_le` needed now?

oscarbenjamin commented on sympy/polys/domains/finitefield.py:
```diff
@@ -57,6 +57,7 @@ def get_field(self):
         return self
 
     def to_sympy(self, a):
+        from sympy.polys.domains.groundtypes import SymPyInteger
```
Is this change still needed?

oscarbenjamin commented on sympy/polys/rootoftools.py:
```diff
@@ -970,42 +971,49 @@ def eval_rational(self, dx=None, dy=None, n=15):
+
+@dispatch(ComplexRootOf, ComplexRootOf)
+def _eval_is_eq(lhs, rhs): # noqa:F811
+    return sympify(lhs == rhs)
```
If `is_eq` uses fuzzy bool then the `sympify` should be removed

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -23,6 +23,11 @@ class Expr(Basic, EvalfMixin):
+    If you want to override the comparisons of expressions:
+    Should use _eval_is_ge for inequality, or _eval_Eq, with multiple dispatch.
+    _eval_is_ge return true if x >= y, false if x < y, and None if the two types
+    cannot be compared.
```
This should say `_eval_is_eq` rather than `_eval_Eq`. Also "if the two types are not comparable or the comparison is indeterminate".

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1049,204 @@ def _eval_relation(cls, lhs, rhs):
+            return retval
+
+    retval = lhs._eval_Eq(rhs)
+    if retval is not None:
```
Is there still an `_eval_Eq` method defined on `Basic`? Maybe it needs to still be there for backward compatibility...

I would want to do `getattr(lhs, '_eval_Eq', None)` to allow the possibility that the input doesn't have an `_eval_Eq` method but I guess that isn't needed if it's defined on `Basic`.

There should be a comment here explaining that `_eval_Eq` is only being called for backwards compatibility reasons and that anyone wanting to overload this should use `_eval_is_eq`.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1049,204 @@ def _eval_relation(cls, lhs, rhs):
+    return fuzzy_not(is_eq(lhs, rhs))
+
+
+def is_eq(lhs, rhs):
```
These top-level functions `is_eq`, `is_ne`, etc should have docstrings with examples. They don't need to be added to the docs but an explanation and a clear tested demonstration would be good. There could be one docstring for `is_eq` and one for `is_ge` and the others can refer to them for more details e.g.:
```
def is_neq(lhs, rhs):
    """Fuzzy bool for lhs does not equal rhs.

    See the docstring for is_eq for more
    """
```

I think it would be good for the docstring for `is_eq` to explain (with an example) that `_eval_is_eq` is used for both `is_eq` and `is_ne`. Likewise the `is_ge` docstring should explain how to overload `_eval_is_ge` and that it covers all of `is_le` etc.

seadavis replied:
I've added these top level functions, but I'm not sure if there is a way to include an example with overload _eval_is_ge etc. when doing the docstring

oscarbenjamin replied:
You could make up a class, something like
```python
def is_ge(lhs, rhs):
    """
    >>> class Power(Expr):
    ...     def __new__(cls, base, exp):
    ...         return Basic.__new__(_sympify(base), _sympify(exp))
    >>> @dispatch(Power, Power):
    ... def _eval_is_ge(lhs, rhs):
    ...     if fuzzy_and(is_ge(lhs.args[0], rhs.args[0]), is_ge(lhs.args[1], rhs.args[1]))
    ...         return True
    >>> Power(2, 3) < Power(3, 6)
    True
    """
```

oscarbenjamin replied:
There should be a more involved example in the tests, including tests that `_eval_Eq` still works.

seadavis replied:
Should the _eval_Eq portion be in the unit tests, or the doctests or both?

oscarbenjamin replied:
That should be the unit tests. The doctests should demonstrate the code we want users to use.

oscarbenjamin commented:
This is looking good - almost there!

seadavis committed:
Documentation, and fixing rootof, and Variable in Ast

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+    Fuzzy boolean used to check if lhs is equal, mathematically
+    to rhs.
+    Returns True or False if the two are comparable, or None if the two are incomparable
+    or if the comparison is indeterminate.
```
The word comparable is tricky because sympy uses it to mean something different:
```julia
In [1]: (1 + sqrt(2)).is_comparable                                                                                                            
Out[1]: True
```
That meaning of comparable is "the expression is real-valued and can be evaluated to a simple `Float`". The sense of comparability is that that makes it usually possible to evaluate an inequality `a<b` by numerical comparison.

How about something like:

Fuzzy bool representing mathematical equality between lhs and rhs.

If lhs and rhs are easily shown to be equal (or unequal) then True (or False) is returned. Otherwise None is returned meaning that lhs and rhs may or may not be equal.

Notes:

This function is intended to give a relatively fast determination and deliberately does not attempt slow calculations that might help in obtaining a determination of True or False in more difficult cases. The evaluation of an `Equality` uses `is_eq` internally so if `is_eq(a, b)` gives True (or False) then `Eq(a, b)` will give `S.true` or `S.false` and vice versa. When `is_eq(a, b)` returns `None`, `Eq(a, b)` will give an unevaluated `Eq`.

Since `is_eq` is primarily an _internal_ function it is expected that the arguments will be sympified already.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+    True
+
+    >>> from sympy.core.relational import is_neq
+    >>> from sympy.core.sympify import _sympify
```
It isn't necessary to reimport a function that has already been imported by a doctest in the same docstring. It's better just to put all imports at the top and then show the examples with some explanation. It's also worth showing how `a == b` and `Eq(a, b)` would behave with similar arguments.

For user documentation we wouldn't normally use `_sympify` rather than `sympify`. In fact `S` is fine e.g. `S(0)`.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+
+    is_le calls is_ge swapping rhs and lhs
+    is_lt calls is_ge and then fuzzy nots the result
+    is_gt calls is_le and fuzzy nots the result
```
Rather than these details it is better just to say that other which functions are based on this one.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+
+def is_ge(lhs, rhs):
+    """
+    Fuzzy bool for lhs is greater than or equal to rhs.
```
This docstring should clarify the behaviour for non-real values as well as for non-Expr or other types like Matrix etc. There are other types in sympy that support `>=` but would not be usable with `is_ge` e.g. sets:
```julia
In [3]: FiniteSet(1, 2) <= FiniteSet(1, 2, 3)                                                                                                  
Out[3]: True
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+    if eval_func is not None:
+        retval = eval_func(lhs)
+        if retval is not None:
+            return retval
```
We could simplify the above duplication using a loop e.g.
```
for side1, side2 in (lhs, rhs), (rhs, lhs):
    eval_func = ...
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,309 @@ def _eval_relation(cls, lhs, rhs):
+            return retval
+
+    eval_func = getattr(lhs, '_eval_Eq', None)
+    if eval_func is not None:
```
There should be a comment here explaining that this method is being called for backwards compatibility and that it is better to use `_eval_is_eq` instead.

seadavis commented:
@oscarbenjamin I appreciate you taking the time to write helpful comments, I am learning alot. I will integrate your suggestions within the next few days, as I have time.

oscarbenjamin commented:
Thanks for the good work! This will be a major improvement in core.

seadavis committed:
Update relational.py

seadavis committed:
Merge branch 'master' into 18778_FuzzyComparisons

seadavis committed:
updated comments added tests

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,287 @@ def _eval_relation(cls, lhs, rhs):
+    is_lt calls is_ge
+
+    InEquality classes, such as Lt, Gt, etc. Use one of is_ge, is_le, etc.
+    If is_ge(a, b) returns nont the Ge(a, b) will just return an unevalutated comparison
```
nont should be None. unevalutated should unevaluated

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,287 @@ def _eval_relation(cls, lhs, rhs):
+    Returns None if lhs cannot be compared to rhs, or if the comparison is indeterminate
+
+    is_le calls is_ge
+    is_lt calls is_ge
```
We could have a full table here like
```
The four comparison functions ``is_le``, ``is_lt``, ``is_ge``, and ``is_gt`` are
each implemented in terms of ``is_ge`` in the following way:

is_ge(x, y) := is_ge(x, y)
is_le(x, y) := is_ge(y, x)
is_lt(x, y) := fuzzy_not(is_ge(x, y))
is_gt(x, y) = fuzzy_not(is_ge(y, x))

To maintain these equivalences in fuzzy logic it is important that in cases where
either x or y is non-real all comparisons will give None.
```

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,287 @@ def _eval_relation(cls, lhs, rhs):
+
+    Returns True if lhs is greater than or equal to rhs
+    Returns False if lhs is not greater than rhs
+    Returns None if lhs cannot be compared to rhs, or if the comparison is indeterminate
```
We should have a full parameters and returns section in the docstring I think. The parameters section should clarify that inputs are expected to be sympified already and should be instances of Expr.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -1138,3 +1047,287 @@ def _eval_relation(cls, lhs, rhs):
+    If is_ge(a, b) returns nont the Ge(a, b) will just return an unevalutated comparison
+
+    so, to override inequality you just need to write an _eval_is_ge
+    using multiple dispatch
```
```
To implement comparisons with ``Gt(a, b)`` or ``a > b`` etc for an ``Expr`` subclass
it is only necessary to define a dispatcher method for ``_eval_is_ge`` like

    >>> class MyExpr(Expr):
    ...     def __new__(cls, arg):
    ...         return Expr.__new__(cls, sympify(arg)) 
    ...     @property
    ...     def value(self):
    ...         return self.args[0]
    ... 
    >>> @dispatch(MyExpr, MyExpr):
    ... def _eval_is_ge(a, b):
    ...     return is_ge(a.value, b.value)
    ...
    >>> a = MyExpr(1)
    >>> b = MyExpr(2)
    >>> a < b
    True
    >>> a <= b
    True
    >>> a > b
    False
    >>> is_lt(a, b)
    False
```

oscarbenjamin commented on sympy/core/tests/test_relational.py:
```diff
@@ -1129,3 +1132,54 @@ def test_multivariate_linear_function_simplification():
+           return Basic.__new__(PowTest, _sympify(base), _sympify(exp))
+
+
+        def _eval_Eq(lhs, rhs):
```
There should only one blank line above.

This test should have a comment clarifying that it is testing `_eval_Eq` for backwards compatibility because the new method is to use `_eval_is_eq`

oscarbenjamin commented on sympy/matrices/immutable.py:
```diff
@@ -178,7 +165,7 @@ def __setitem__(self, *args):
         return SparseMatrix.__getitem__(self, (i, j))
 
-    _eval_Eq = ImmutableDenseMatrix._eval_Eq
+    # _eval_Eq = ImmutableDenseMatrix._eval_Eq
```
This should be removed if it is not needed any more

oscarbenjamin commented on sympy/sets/__init__.py:
```diff
@@ -1,13 +1,14 @@
 from .powerset import PowerSet
 from ..core.singleton import S
-
+from .handlers.comparison import _eval_is_eq  # noqa:F401
```
Why is this import needed?

seadavis replied:
Because without it, the dispatcher did not find _eval_is_eq for the sets when they were in the handlers file

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -924,7 +915,7 @@ def __new__(cls, start, end, left_open=False, right_open=False):
 
         # evaluate if possible
-        if (end < start) == True:
+        if (end >= start) == False:
```
This kind of code `if (end < start) == True` is really a workaround for not having `is_ge`. This should really be
```
if is_lt(end, start):
```

oscarbenjamin commented:
I had a few comments but this is pretty much ready to be merged. The tests have failed due to #19700 

Right now II just want to get this finished and merged.

Afterwards I think there could followup PRs to change a couple of things:

1. I think it's best to use multiple dispatch with the function name `_` as in #19463. This avoids the `noqa` it also means that to add a dispatch routine you don't need to import `dispatch` from multipledispatch.

2. I think maybe the logic going from
```
__le__ -> Le -> is_le -> Le(evaluate=False)
```
can be shortcut. It would be better for `__le__` to call `is_le` and either return the result or call `Le(evaluate=False)` directly.

seadavis committed:
Updated documentation, and sets comparison

seadavis committed:
Update relational.py

seadavis commented:
@oscarbenjamin ok, sounds good. I believe I have now taken care of all of your comments.

oscarbenjamin commented:
Looks good. The tests will fail again due to #19700

oscarbenjamin commented:
This is good. I'll merge it now.

Thanks for this!
